File: DataStructures/Bit mask/leetcode/37. Sudoku Solver/Sudoku-Solver-Claude.py
# Runtime
# 1808
# ms
# Beats
# 52.41%
# Memory
# 17.92
# MB
# Beats
# 62.04%

# Time Limit Exceededが発生したのは、最も効率的でないセル探索順序が原因です。空のセルを見つけるたびに最初から探索するのではなく、より効率的な最適化を実装します。Time Limit Exceededを解決するために、以下の最適化を実装しました：

# **主要な最適化ポイント：**

# 1. **事前計算によるセット管理**
#    - 各行・列・ボックスで使用済みの数字をSetで管理
#    - 毎回全体をチェックせず、O(1)での制約確認

# 2. **空セルの事前収集**
#    - 最初に全ての空セルを収集してリスト化
#    - 毎回探索する必要がなく、直接インデックスアクセス

# 3. **使用可能数字の効率計算**
#    - セット演算で使用可能な数字を一度に計算
#    - 不要なループを削減

# 4. **メモリ効率の向上**
#    - セットの追加/削除でO(1)操作
#    - 重複チェックが高速化

# **計算量改善：**
# - **従来**: 各セルで9×9の全体チェック → O(81)
# - **最適化後**: セット操作での制約チェック → O(1)

# **特に効果的な場面：**
# - 空セルが多いケース（提供されたテストケースのような）
# - 制約が複雑に絡み合うケース

# この実装により、Time Limit Exceededを回避しつつ、LeetCodeの厳しい時間制限内での実行が可能になります。

# from typing import List, Optional, Tuple, Set
from typing import List, Tuple, Set

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        数独パズルを解く関数（boardを直接変更する）
        
        Args:
            board: 9x9の数独ボード（'.'は空のセル、'1'-'9'は既に配置された数字）
        
        Returns:
            None（boardを直接変更する）
        """
        # 制約セットを事前計算
        rows: List[Set[str]] = [set() for _ in range(9)]
        cols: List[Set[str]] = [set() for _ in range(9)]
        boxes: List[Set[str]] = [set() for _ in range(9)]
        empty_cells: List[Tuple[int, int]] = []
        
        # 初期状態を設定
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    empty_cells.append((i, j))
                else:
                    val = board[i][j]
                    rows[i].add(val)
                    cols[j].add(val)
                    boxes[self._get_box_index(i, j)].add(val)
        
        self._solve_optimized(board, empty_cells, 0, rows, cols, boxes)
    
    def _solve_optimized(self, board: List[List[str]], empty_cells: List[Tuple[int, int]], 
                        idx: int, rows: List[Set[str]], cols: List[Set[str]], 
                        boxes: List[Set[str]]) -> bool:
        """
        最適化されたバックトラッキング解法
        
        Args:
            board: 9x9の数独ボード
            empty_cells: 空のセルのリスト
            idx: 現在処理中のセルインデックス
            rows: 各行で使用済みの数字セット
            cols: 各列で使用済みの数字セット
            boxes: 各ボックスで使用済みの数字セット
            
        Returns:
            bool: 解けた場合はTrue、解けない場合はFalse
        """
        if idx == len(empty_cells):
            return True  # すべての空セルが埋まった
        
        row, col = empty_cells[idx]
        box_idx = self._get_box_index(row, col)
        
        # 使用可能な数字を事前計算
        used = rows[row] | cols[col] | boxes[box_idx]
        available = {'1', '2', '3', '4', '5', '6', '7', '8', '9'} - used
        
        for num in available:
            # 数字を配置
            board[row][col] = num
            rows[row].add(num)
            cols[col].add(num)
            boxes[box_idx].add(num)
            
            # 再帰的に次のセルを処理
            if self._solve_optimized(board, empty_cells, idx + 1, rows, cols, boxes):
                return True
            
            # バックトラック
            board[row][col] = '.'
            rows[row].remove(num)
            cols[col].remove(num)
            boxes[box_idx].remove(num)
        
        return False
    
    def _get_box_index(self, row: int, col: int) -> int:
        """
        指定された座標の3x3ボックスインデックスを取得
        
        Args:
            row: 行インデックス
            col: 列インデックス
            
        Returns:
            int: ボックスインデックス（0-8）
        """
        return (row // 3) * 3 + (col // 3)

File: DataStructures/Bit mask/leetcode/37. Sudoku Solver/Sudoku-Solver-Claude.ts
// Runtime
// 445
// ms
// Beats
// 60.35%
// Memory
// 56.69
// MB
// Beats
// 84.48%

// 数独パズルを解くTypeScript実装を提供しました。この解法の特徴：

// **アルゴリズム：バックトラッキング**
// - 空のセル（'.'）を順番に見つけて、1-9の数字を試行
// - 各数字が配置可能かルールをチェック
// - 配置できない場合は前の選択に戻る（バックトラック）

// **最適化のポイント：**
// 1. **メモリ効率**: 元のboardを直接変更し、新しい配列を作成しない
// 2. **処理速度**: 最初に見つけた空のセルから処理することで、不要な探索を避ける
// 3. **早期終了**: 無効な配置が見つかった時点で即座にfalseを返す

// **時間複雑度**: O(9^(空のセル数)) - 最悪の場合
// **空間複雑度**: O(1) - 追加のメモリを使用しない（再帰スタックを除く）

// この実装は、制約条件（各行・列・3x3ボックスに1-9が一度ずつ）を効率的にチェックし、LeetCodeの時間・メモリ制限内で動作するよう設計されています。

/**
 * 数独パズルを解く関数
 * @param board - 9x9の数独ボード（'.'は空のセル、'1'-'9'は既に配置された数字）
 * @returns void（boardを直接変更する）
 */
function solveSudoku(board: string[][]): void {
    solve(board);
}

/**
 * バックトラッキングを使用して数独を解く再帰関数
 * @param board - 9x9の数独ボード
 * @returns boolean - 解けた場合はtrue、解けない場合はfalse
 */
function solve(board: string[][]): boolean {
    // 空のセルを探す
    for (let row = 0; row < 9; row++) {
        for (let col = 0; col < 9; col++) {
            if (board[row][col] === '.') {
                // 1-9の数字を試す
                for (let num = 1; num <= 9; num++) {
                    const char = num.toString();
                    
                    // その数字が配置可能かチェック
                    if (isValid(board, row, col, char)) {
                        board[row][col] = char;
                        
                        // 再帰的に次のセルを試す
                        if (solve(board)) {
                            return true;
                        }
                        
                        // バックトラック：配置を取り消す
                        board[row][col] = '.';
                    }
                }
                return false; // どの数字も配置できない場合
            }
        }
    }
    return true; // すべてのセルが埋まった場合
}

/**
 * 指定された位置に数字を配置できるかチェックする関数
 * @param board - 9x9の数独ボード
 * @param row - 行のインデックス（0-8）
 * @param col - 列のインデックス（0-8）
 * @param num - チェックする数字（文字列）
 * @returns boolean - 配置可能な場合はtrue、不可能な場合はfalse
 */
function isValid(board: string[][], row: number, col: number, num: string): boolean {
    // 行のチェック
    for (let x = 0; x < 9; x++) {
        if (board[row][x] === num) {
            return false;
        }
    }
    
    // 列のチェック
    for (let x = 0; x < 9; x++) {
        if (board[x][col] === num) {
            return false;
        }
    }
    
    // 3x3サブボックスのチェック
    const startRow = Math.floor(row / 3) * 3;
    const startCol = Math.floor(col / 3) * 3;
    
    for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
            if (board[startRow + i][startCol + j] === num) {
                return false;
            }
        }
    }
    
    return true;
}